### ISSUE TITLE
Question: How much memory does ghc-mod consume when editing Futhark source?

### ISSUE THREAD

**Issue by mrakgr** [HIPERFIT/futhark#188]
I am still doing Haskell exercises, but as a preliminary I am currently trying to get editor services to work with Futhark. I figured out how to get the missing dependencies going with Stack by upgrading it to the latest version, but now I am running into memory consumption issues with `ghc-mod`. 

Inside the Atom editor, it draws in 5Gb out of 8Gb on my machine and then runs out of memory. I do not suppose you could check how much memory online type-checking takes up inside the task manager or equivalent on Linux?

I'd like to figure this one out otherwise I am going to have to do my coding on Futhark without help from the editor, which is something I would very much like to avoid. Currently, I am [pestering](https://github.com/atom-haskell/haskell-ghc-mod/issues/149#issuecomment-221247439) lierdakil on his page.

**Comment by Athas** [comment]
I don't use ghc-mod myself (although I used to), but I have an idea of
what may be wrong.  When using 'stack repl' to load the entire Futhark
compiler into ghci, the memory usage also goes through the roof.  If
ghc-mod does something like that, which is reasonable if it wants to do
introspection, it might run into the same problem.

I'm not completely certain why this happens.  Sure, the Futhark codebase
is not trivial, but there are plenty of bigger Haskell programs.  I
suspect that it has something to do with the parser, which is generated
from src/Language/Futhark/Parser/Parser.y by the parser generator
'happy'.  I have previously noticed ghci going haywire when loading it,
presumably because whatever code structure is produced hits a worst-case
in ghci.

-- 
\  Troels
/\ Henriksen

**Comment by mrakgr** [comment]
I've had no luck with disabling autocomplete. When I try to load pyopencl using `stack repl` it takes up 2.35Gb.

**Comment by mrakgr** [comment]
Does 2.35Gb seem like more than usual to you or is it standard?

**Comment by Athas** [comment]
The ghci I currently have open (which has not loaded the parser) is
using 2.7GiB of resident memory.

Using 'stack repl' to load futhark-pyopencl results in a process that
uses 3.7GiB.

This is with GHC 8.0.  I vaguely recall GHC 7.10 using significantly
more.

-- 
\  Troels
/\ Henriksen

**Comment by mrakgr** [comment]
Hmmmm...I have no idea. It is more than on my computer at any rate.

No matter what, even if I decide not to use `ghc-mod`, I am going to have to find a way to get some real time feedback. Maybe I'll set up a shortcut so the program compiles every time I save. Right now, GHC is doing type checking every time I do that.

Last time I asked you, you said you were using Emacs and now you said you aren't using `ghc-mod` which is necessary to give Haskell modern IDE features. I am not sure whether I should be impressed or horrified. Personally, I really like the modern IDE stuff - I've noticed that it significantly impacts my productivity in a positive fashion. And now that I've used Haskell for a bit, I actually think that F# has better type inference than it, contrary to some of the things that I've read. Not necessarily while the program is working, but while it is being worked on.

For example, I would declare a new variable inside a let statement, save and the information on past types would get lost in Haskell. It is one of the things I had to get used to while moving to Haskell.

I used Emacs a year ago for a little bit, while I was doing the Compilers course, but found it strange at the time and did not really make a significant effort into learning it. I switched to the Code::Blocks IDE for the later assignments. I do not suppose it has some features that would make up for the lack of `ghc-mod`? I have no idea how I would live if I was forced to program without feedback from the IDE.

**Comment by Athas** [comment]
Yeah, the Haskell tooling isn't as good as it could be.  That said, I
believe F# has the best tooling of _any_ functional language, so it's
hard to beat.  Have you looked at Leksah?  It's a Haskell IDE that seems
to work like ordinary IDEs, but I have never tried it myself.

My haskell-mode setup isn't completely dark-ages. :-) I use ghci to test
incrementally, and Emacs automatically runs ctags, which allows me to
use C-dot to jump to definitions (most of the time).  You can also setup
flymake, which transparently recompiles in the background and underlines
errors with red squigglies.

It is still a little crude and requires manual assembly,
unfortunately. I do miss my old Lisp programming setup.

-- 
\  Troels
/\ Henriksen

**Comment by mrakgr** [comment]
I tried Leksah when I heard about Haskell for the first time after I finished the Scala Coursera course which was a month or two after I'd finished the Compilers course. I was doing a lot of online courses at a time to get into shape after not doing any programming for nearly a decade.

At any rate, I could not get it to run. I think there were some problems with the GHC installation or something like that, but at that time I could not get even the Atom plugins that I have now to run. Running stuff from the command line was brand new to me as well. Leksah would silently fail and then crash, so I abandoned it. At any rate, I actually quite like Atom, but it does not really matter as even if I used Leksah, it would be using `ghc-mod` under the hood and I would have the same problems as with Atom.

This thing really derailed me today. I think I'll give it a day or two and then open an issue on the `ghc-mod` page and see where that leads. Maybe I'll look into finding something like `flymake` for Atom.

I did what I can for now. I'll get back to working on the problem that was doing. 

It was nice talking to you.
